lib/Core/AddressSpace.h

```diff
@@ -74,20 +74,35 @@ namespace klee {
                     TimingSolver *solver,
                     ref<Expr> address,
                     ObjectPair &result,
-                    bool &success);
+                    bool &success) const;
 
-    /// Resolve address to a list of ObjectPairs it can point to. If
-    /// maxResolutions is non-zero then no more than that many pairs
-    /// will be returned. 
+    /// Resolve pointer `p` to a list of `ObjectPairs` it can point
+    /// to. If `maxResolutions` is non-zero then no more than that many
+    /// pairs will be returned.
     ///
-    /// \return true iff the resolution is incomplete (maxResolutions
-    /// is non-zero and the search terminated early, or a query timed out).
+    /// \return true iff the resolution is incomplete (`maxResolutions`
+    /// is non-zero and it was reached, or a query timed out).
     bool resolve(ExecutionState &state,
                  TimingSolver *solver,
-                 ref<Expr> address, 
+                 ref<Expr> p,
                  ResolutionList &rl, 
                  unsigned maxResolutions=0,
-                 double timeout=0.);
+                 double timeout=0.) const;
+
+
+    /// Check if pointer `p` can point to the memory object in the
+    /// given object pair.  If so, add it to the given resolution list.
+    ///
+    /// \return 1 iff the resolution is incomplete (`maxResolutions`
+    /// is non-zero and it was reached, or a query timed out), 0 iff
+    /// the resolution is complete (`p` can only point to the given
+    /// memory object), and 2 otherwise.
+    int checkPointerInObject(ExecutionState &state,
```

Comment:
I don't think this is the right refactoring. There should be two functions

`bool isPointerInObject(state,solver,p, MemoryObject o)`
and
`bool pointerMustBeInObject(state,solver,p, MemoryObject o)`.
that of course mean that you duplicate some `ResolutionList` handling logic.

But in the current version you are passing in this `ObjectPair`, which you only need so you can add it to the `ResolutionList, both of them have nothing to do with checking if a pointer is in object.

The two functions I suggested would actually have use outside of this refactoring and make sense as standalone functions. The current function is only useful for removing code duplication.

Reply:
I'm not saying that it's the best refactoring, but the main goal this is trying to achieve is to get rid of the non-deterministic coverage which makes one of the Codecov occasionally fail.  This refactoring ensures that; if you separate this into multiple functions, you might have the same issue again.  I also think it's more readable than what we had before, but this is not why I did this in the first place.  Of course, I realize this is fragile, but it does solve the longstanding issue of failed checks.

I think we should merge this right away, or disable the Codecov checks.

Reply:
Good point; no reason to be part of the API.  I'll make it `private`.